File: app/src/main/java/com/mainstreetcode/teammate/model/enums/StatType.java
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.model.enums;

import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonObject;
import com.google.gson.JsonSerializationContext;
import com.tunjid.androidbootstrap.recyclerview.diff.Differentiable;
import com.mainstreetcode.teammate.util.ModelUtils;
import com.tunjid.androidbootstrap.core.text.SpanBuilder;

import java.util.Objects;

public class StatType extends MetaData {

    private String emoji;
    private String sportCode;
    private StatAttributes attributes = new StatAttributes();

    private StatType(String code, String name, String emoji, String sportCode) {
        super(code, name);
        this.emoji = emoji;
        this.sportCode = sportCode;
    }

    public static StatType empty() {
        return new StatType("", "", "", "");
    }

    public CharSequence getEmoji() { return ModelUtils.processString(emoji); }

    public CharSequence getEmojiAndName() {
        return SpanBuilder.of(getEmoji()).append("   ").append(name).build();
    }

    public StatAttributes getAttributes() {
        return attributes;
    }

    public StatAttribute fromCode(String code) {
        for (StatAttribute attr : attributes) if (attr.code.equals(code)) return attr;
        return StatAttribute.empty();
    }

    @Override
    public String toString() {
        return name;
    }

    public void update(StatType updated) {
        super.update(updated);
        this.emoji = updated.emoji;
        this.sportCode = updated.sportCode;
        ModelUtils.replaceList(attributes, updated.attributes);
    }

    @Override
    public boolean areContentsTheSame(Differentiable other) {
        boolean result = super.areContentsTheSame(other);
        if (!(other instanceof StatType)) return result;
        return result && sportCode.equals(((StatType) other).sportCode);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof StatType)) return false;
        StatType variant = (StatType) o;
        return Objects.equals(code, variant.code) && Objects.equals(name, variant.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(code, name);
    }

    public static class GsonAdapter extends MetaData.GsonAdapter<StatType> {

        private static final String EMOJI_KEY = "emoji";
        private static final String SPORT = "sport";
        private static final String ATTRIBUTES = "attributes";

        @Override
        StatType fromJson(String code, String name, JsonObject body, JsonDeserializationContext context) {
            String emoji = ModelUtils.asString(EMOJI_KEY, body);
            String sportCode = ModelUtils.asString(SPORT, body);

            StatType type = new StatType(code, name, emoji, sportCode);
            ModelUtils.deserializeList(context, body.get(ATTRIBUTES), type.attributes, StatAttribute.class);
            return type;
        }

        @Override
        JsonObject toJson(JsonObject serialized, StatType src, JsonSerializationContext context) {
            serialized.addProperty(EMOJI_KEY, src.emoji);
            serialized.addProperty(SPORT, src.sportCode);
            serialized.add(ATTRIBUTES, context.serialize(src.attributes));

            return serialized;
        }
    }
}

File: app/src/main/java/com/mainstreetcode/teammate/model/enums/StatType.kt
/*
 * MIT License
 *
 * Copyright (c) 2019 Adetunji Dahunsi
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package com.mainstreetcode.teammate.model.enums

import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonObject
import com.google.gson.JsonSerializationContext
import com.mainstreetcode.teammate.util.asStringOrEmpty
import com.mainstreetcode.teammate.util.deserializeList
import com.mainstreetcode.teammate.util.processEmoji
import com.mainstreetcode.teammate.util.replace
import com.tunjid.androidbootstrap.core.text.SpanBuilder
import com.tunjid.androidbootstrap.recyclerview.diff.Differentiable
import java.util.*

class StatType private constructor(
        code: String,
        name: String,
        emoji: CharSequence,
        private var sportCode: String
) : MetaData(code, name) {

    val attributes: StatAttributes = StatAttributes()

    var emoji: CharSequence = emoji
        get() = field.processEmoji()

    val emojiAndName: CharSequence
        get() = SpanBuilder.of(emoji).append("   ").append(name).build()

    fun fromCode(code: String): StatAttribute {
        for (attr in attributes) if (attr.code == code) return attr
        return StatAttribute.empty()
    }

    override fun toString(): String = name

    fun update(updated: StatType) {
        super.update(updated)
        this.emoji = updated.emoji
        this.sportCode = updated.sportCode
        attributes.replace(updated.attributes)
    }

    override fun areContentsTheSame(other: Differentiable): Boolean {
        val result = super.areContentsTheSame(other)
        return if (other !is StatType) result else result && sportCode == other.sportCode
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is StatType) return false
        val variant = other as StatType?
        return code == variant!!.code && name == variant.name
    }

    override fun hashCode(): Int = Objects.hash(code, name)

    class GsonAdapter : MetaData.GsonAdapter<StatType>() {

        override fun fromJson(code: String, name: String, body: JsonObject, context: JsonDeserializationContext): StatType {
            val emoji = body.asStringOrEmpty(EMOJI_KEY)
            val sportCode = body.asStringOrEmpty(SPORT)

            val type = StatType(code, name, emoji, sportCode)
            context.deserializeList(body.get(ATTRIBUTES), type.attributes, StatAttribute::class.java)
            return type
        }

        override fun toJson(serialized: JsonObject, src: StatType, context: JsonSerializationContext): JsonObject {
            serialized.addProperty(EMOJI_KEY, src.emoji.toString())
            serialized.addProperty(SPORT, src.sportCode)
            serialized.add(ATTRIBUTES, context.serialize(src.attributes))

            return serialized
        }

        companion object {

            private const val EMOJI_KEY = "emoji"
            private const val SPORT = "sport"
            private const val ATTRIBUTES = "attributes"
        }
    }

    companion object {

        fun empty(): StatType = StatType("", "", "", "")
    }
}
